src/ArborX_LinearBVH.hpp

```diff
@@ -172,22 +170,52 @@ class BoundingVolumeHierarchy
              Experimental::TraversalPolicy const &policy =
                  Experimental::TraversalPolicy()) const
   {
+    Details::check_valid_callback<int>(callback, predicates);
     base_type::query(space, predicates,
                      Details::LegacyCallbackWrapper<Callback>{callback},
                      policy);
   }
 
-  template <typename ExecutionSpace, typename Predicates,
-            typename CallbackOrView, typename View, typename... Args>
+  template <typename ExecutionSpace, typename Predicates, typename View,
+            typename... Args>
   std::enable_if_t<Kokkos::is_view_v<std::decay_t<View>>>
-  query(ExecutionSpace const &space, Predicates const &predicates,
-        CallbackOrView &&callback_or_view, View &&view, Args &&...args) const
+  query(ExecutionSpace const &space, Predicates const &predicates, View &&view,
+        Args &&...args) const
   {
-    base_type::query(space, predicates,
-                     std::forward<CallbackOrView>(callback_or_view),
+    base_type::query(space, predicates, Details::LegacyDefaultCallback{},
                      std::forward<View>(view), std::forward<Args>(args)...);
   }
 
+  template <typename ExecutionSpace, typename Predicates, typename Callback,
+            typename OutputView, typename OffsetView, typename... Args>
+  std::enable_if_t<!Kokkos::is_view_v<std::decay_t<Callback>>>
+  query(ExecutionSpace const &space, Predicates const &predicates,
+        Callback const &callback, OutputView &&out, OffsetView &&offset,
+        Args &&...args) const
+  {
+    if constexpr (!Details::is_tagged_post_callback<
+                      std::decay_t<Callback>>::value)
+    {
+      Details::check_valid_callback<int>(callback, predicates, out);
+      base_type::query(
+          space, predicates, Details::LegacyCallbackWrapper<Callback>{callback},
```

Comment:
I think your real issue was here.  Were you getting memory access violations at runtime?  You'd need to decay the template parameter

Reply:
Why did you go for dropping the std::decay instead of forwarding the callback and "fixing" the template argument?

Reply:
OK, I'll try. But can you explain what we gain by forwarding the callback?

Reply:
One less copy when passing a temporary (which doesn't really matter)
I mostly wanted to make sure we understand what was wrong with the initial version.
If you prefer const ref I will approve but would you please try locally so we know for sure.

Reply:
OK, you are right. I updated to decay the callback template argument and it worked.